Libraries/Blob/FileReader.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule FileReader
+ * @flow
+ */
+
+'use strict';
+
+const EventTarget = require('event-target-shim');
+const Blob = require('Blob');
+const { FileReaderModule } = require('NativeModules');
+
+type ReadyState =
+  | 0 // EMPTY
+  | 1 // LOADING
+  | 2 // DONE
+
+type ReaderResult = string | ArrayBuffer;
+
+const READER_EVENTS = [
+  'abort',
+  'error',
+  'load',
+  'loadstart',
+  'loadend',
+  'progress',
+];
+
+const EMPTY = 0;
+const LOADING = 1;
+const DONE = 2;
+
+class FileReader extends EventTarget(...READER_EVENTS) {
+
+  static EMPTY = EMPTY;
+  static LOADING = LOADING;
+  static DONE = DONE;
+
+  EMPTY = EMPTY;
+  LOADING = LOADING;
+  DONE = DONE;
+
+  _readyState: ReadyState;
+  _error: ?Error;
+  _result: ?ReaderResult;
+  _aborted: boolean = false;
+  _subscriptions: Array<*> = [];
+
+  constructor() {
+    super();
+    this._reset();
+  }
+
+  _reset(): void {
+    this._readyState = EMPTY;
+    this._error = null;
+    this._result = null;
+  }
+
+  _clearSubscriptions(): void {
+    this._subscriptions.forEach(sub => sub.remove());
+    this._subscriptions = [];
+  }
+
+  _setReadyState(newState: ReadyState) {
+    this._readyState = newState;
+    this.dispatchEvent({ type: 'readystatechange' });
+    if (newState === DONE) {
+      if (this._aborted) {
+        this.dispatchEvent({ type: 'abort' });
+      } else if (this._error) {
+        this.dispatchEvent({ type: 'error' });
+      } else {
+        this.dispatchEvent({ type: 'load' });
+      }
+      this.dispatchEvent({ type: 'loadend' });
+    }
+  }
+
+  readAsArrayBuffer() {
+    throw new Error('FileReader.readAsArrayBuffer is not implemented');
+  }
+
+  readAsDataURL(blob: Blob) {
+    this._aborted = false;
+
+    FileReaderModule.readAsDataURL(blob.data).then((text: string) => {
```

Comment:
Can we make it async? Looks like a potential perf bottleneck

Reply:
What do you mean by that? Isn't already async?

Reply:
Sorry, for some reason thought you return result of this call. No idea why 😐